src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketBuffer.cs

```diff
@@ -0,0 +1,694 @@
+
+            Buffer.BlockCopy(payload.Array,
+                payload.Offset + unconsumedDataOffset,
+                _payloadBuffer.Array,
+                _payloadBuffer.Offset,
+                bytesBuffered);
+
+            _bufferedPayloadReceiveResult =
+                new WebSocketReceiveResult(bytesBuffered, messageType, endOfMessage);
+
+            this.ValidateBufferedPayload();
+        }
+
+        internal bool ReceiveFromBufferedPayload(ArraySegment<byte> buffer, out WebSocketReceiveResult receiveResult)
+        {
+            ThrowIfDisposed();
+            ValidateBufferedPayload();
+
+            int bytesTransferred = Math.Min(buffer.Count, _bufferedPayloadReceiveResult.Count);
+
+            receiveResult = new WebSocketReceiveResult(
+                bytesTransferred,
+                _bufferedPayloadReceiveResult.MessageType,
+                bytesTransferred == 0 && _bufferedPayloadReceiveResult.EndOfMessage,
+                _bufferedPayloadReceiveResult.CloseStatus,
+                _bufferedPayloadReceiveResult.CloseStatusDescription);
+
+            Buffer.BlockCopy(_payloadBuffer.Array,
+                _payloadBuffer.Offset + _payloadOffset,
+                buffer.Array,
+                buffer.Offset,
+                bytesTransferred);
+
+            bool morePayloadBuffered;
+            if (_bufferedPayloadReceiveResult.Count == 0)
+            {
+                _payloadOffset = 0;
+                _bufferedPayloadReceiveResult = null;
+                morePayloadBuffered = false;
+            }
+            else
+            {
+                _payloadOffset += bytesTransferred;
+                morePayloadBuffered = true;
+                this.ValidateBufferedPayload();
+            }
+
+            return morePayloadBuffered;
+        }
+
+        internal ArraySegment<byte> ConvertNativeBuffer(WebSocketProtocolComponent.Action action,
+            WebSocketProtocolComponent.Buffer buffer,
+            WebSocketProtocolComponent.BufferType bufferType)
+        {
+            ThrowIfDisposed();
+
+            IntPtr bufferData;
+            uint bufferLength;
+
+            UnwrapWebSocketBuffer(buffer, bufferType, out bufferData, out bufferLength);
+
+            if (bufferData == IntPtr.Zero)
+            {
+                return WebSocketHelpers.EmptyPayload;
+            }
+
+            if (this.IsNativeBuffer(bufferData, bufferLength))
+            {
+                return new ArraySegment<byte>(_internalBuffer.Array,
+                    this.GetOffset(bufferData),
+                    (int)bufferLength);
+            }
+
+            Debug.Assert(false, "'buffer' MUST reference a memory segment within the pinned InternalBuffer.");
+            // Indicates a violation in the contract with native Websocket.dll and could indicate 
+            // memory corruption because the internal buffer is shared between managed and native code
+            throw new AccessViolationException();
```

Comment:
This is a behavior change from desktop, and not a nit. Can you elaborate more on what exception should be thrown here? Please file issue